### ISSUE TITLE
Potentially OS-specific `bcftools.mpileup()` error: `Segmentation fault (core dumped)`

### ISSUE THREAD

**Issue by sbslee** [pysam-developers/pysam#1081]
Hello developers of `pysam`,

As always, thanks for your hard work! Today, I'm reporting a potential OS-specific bug in `bcftools.mpileup()` where the method keeps killing the running program with `Segmentation fault (core dumped)`. The simplest example to reproduce the error is:

```
$ cat pysam_test.py
from pysam import bcftools
bcftools.mpileup('in.bam', '-f', 'ref.fa')
```

```
$ python3 pysam_test.py
Segmentation fault (core dumped)
```

In the example above, I put `'in.bam'` and `'ref.fa'` instead of actual filenames because I've tried a number of different files and they all gave me the same error.

Interestingly, this error only occurs in linux server; it works just fine in macOS. That's why I think this might be a OS-specific bug. Below is my linux server:

```
$ uname -a
Linux bdcm03 3.10.0-514.2.2.el7.x86_64 #1 SMP Tue Dec 6 23:06:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

And my macOS:

```
$ uname -a
Darwin Seung-beens-Mac-mini.local 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:06:51 PST 2021; root:xnu-7195.81.3~1/RELEASE_ARM64_T8101 arm64
```

I'm using the latest version for both OS systems (0.18.0 for `pysam` and 1.14 for `htslib`, downloaded via conda). I checked other issues in this repository, including my previous ones (#928, #944), but could not find any answers.

Any help would be greatly appreciated!

**Comment by sbslee** [comment]
I did further investigation and it appears the issue is with this [line](https://github.com/pysam-developers/pysam/blob/482e887e67ce035e8a9e8fe1ec55be6c80d3e5dc/pysam/utils.py#L56), which calls `pysam.libcutils._pysam_dispatch()`:

```
$ cat pysam_test.py
from pysam import bcftools
import faulthandler
faulthandler.enable()
bcftools.mpileup('in.bam', '-f', 'ref.fa')
```

```
$ python pysam_test.py 
Fatal Python error: Segmentation fault

Current thread 0x00007fd351a4b740 (most recent call first):
  File "/home/sbslee/anaconda3/envs/fuc/lib/python3.9/site-packages/pysam/utils.py", line 56 in __call__
  File "/mnt/mone/PriFiles/sbslee/Pharmacogenomics/GeT-RM/pysam_test.py", line 4 in <module>
Segmentation fault (core dumped)
```

Any thoughts?

**Comment by sbslee** [comment]
Listing some potentially relevant SO articles regarding why the error appears to occur only with linux server, and not with macOS:

- [Segmentation fault on linux (g++) but not on Mac OS.?](https://stackoverflow.com/questions/37777834/segmentation-fault-on-linux-g-but-not-on-mac-os)
- [Segmentation Fault on Ubuntu Machine but not on Mac](https://stackoverflow.com/questions/20389714/segmentation-fault-on-ubuntu-machine-but-not-on-mac)
- [c - Segmentation Fault on Ubuntu 18.04 but not on macOS High Sierra SDL1.2](https://stackoverflow.com/questions/50107930/c-segmentation-fault-on-ubuntu-18-04-but-not-on-macos-high-sierra-sdl1-2)
- [Segmentation fault happened in Linux but works in Mac](https://stackoverflow.com/questions/32564233/segmentation-fault-happened-in-linux-but-works-in-mac)
- [Segfault in Linux, works fine on Mac OS](https://stackoverflow.com/questions/32835448/segfault-in-linux-works-fine-on-mac-os)

**Comment by jmarshall** [comment]
I am unable to reproduce this with pysam 0.18.0 on Linux.

Please add the following to your test script to verify that you are not accidentally exercising the same old bug in the older version:

```python
import pysam
print("version {} from {}".format(pysam.__version__, pysam))
```

Note that at present Conda gives you the faulty 0.16.0.1 by default; you need to use `conda create -n myenv pysam=0.18.0` to get the up-to-date version of pysam. I will look into this instance of this well-known conda limitation.

**Comment by sbslee** [comment]
@jmarshall,

Thanks so much for your prompt reply! I can confirm the issue is with the latest version (0.18.0):

```
$ cat pysam_test.py
import pysam
from pysam import bcftools
import faulthandler
print("version {} from {}".format(pysam.__version__, pysam))
faulthandler.enable()
bcftools.mpileup('in.bam', '-f', 'ref.fa')
```

```
$ python pysam_test.py 
version 0.18.0 from <module 'pysam' from '/home/sbslee/anaconda3/envs/fuc/lib/python3.9/site-packages/pysam/__init__.py'>
Fatal Python error: Segmentation fault

Current thread 0x00007f318ed55740 (most recent call first):
  File "/home/sbslee/anaconda3/envs/fuc/lib/python3.9/site-packages/pysam/utils.py", line 56 in __call__
  File "/mnt/mone/PriFiles/sbslee/Pharmacogenomics/GeT-RM/pysam_test.py", line 6 in <module>
Segmentation fault (core dumped)
```

**Comment by sbslee** [comment]
P.S. I'd also like to add: To make sure the issue is not caused by any potential pollution in the conda environment (e.g. dependency conflicts), I also repeated this test in a fresh environment installed with:

```
$ conda create -n test pysam=0.18.0
```

**Comment by jmarshall** [comment]
Please show us the output from

```
LD_DEBUG=libs python pysam_test.py 2>&1 | grep libc...tools
```

Do you have any `…PATH` environment variables set? In particular, LD_LIBRARY_PATH or PYTHONPATH?

**Comment by sbslee** [comment]
Attached is the output from your command:

```
(test) [sbslee@bdcm03 GeT-RM]$ LD_DEBUG=libs python pysam_test.py 2>&1 | grep libc...tools
    140333:	find library=libcsamtools.cpython-39-x86_64-linux-gnu.so [0]; searching
    140333:	  trying file=/home/sbslee/anaconda3/envs/test/lib/libcsamtools.cpython-39-x86_64-linux-gnu.so
    140333:	  trying file=/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcsamtools.cpython-39-x86_64-linux-gnu.so
    140333:	find library=libcbcftools.cpython-39-x86_64-linux-gnu.so [0]; searching
    140333:	  trying file=/home/sbslee/anaconda3/envs/test/lib/libcbcftools.cpython-39-x86_64-linux-gnu.so
    140333:	  trying file=/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcbcftools.cpython-39-x86_64-linux-gnu.so
    140333:	calling init: /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcbcftools.cpython-39-x86_64-linux-gnu.so
    140333:	calling init: /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcsamtools.cpython-39-x86_64-linux-gnu.so
```

As for environment variables, I didn't change any from the fresh conda environment with just `pysam=0.18.0`:

```
(test) [sbslee@bdcm03 GeT-RM]$ echo $LD_LIBRARY_PATH
/mnt/mone/ge2011.11/lib/linux-x64
(test) [sbslee@bdcm03 GeT-RM]$ echo $PYTHONPATH

(test) [sbslee@bdcm03 GeT-RM]$
```

**Comment by sbslee** [comment]
P.S. I'm not sure if it's relevant at all, but the latest version of bcftools (1.14) works fine on the same FASTA and BAM files I used in `pysam_test.py`.

**Comment by jmarshall** [comment]
It's grasping at straws, but what is the output from

```
md5sum /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libc{utils,samtools,bcftools}*
```

**Comment by sbslee** [comment]
Oh no, are we already at that stage? :(

Here's the output:

```
(test) [sbslee@bdcm03 GeT-RM]$ md5sum /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libc{utils,samtools,bcftools}*
200b3a84e14d88d8e69194ef1429cd12  /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcutils.cpython-39-x86_64-linux-gnu.so
8a907487f5a653e9409a1c66d0772d6e  /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcutils.pxd
f3617094dc72ea047ad4a30b4698f87d  /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcsamtools.cpython-39-x86_64-linux-gnu.so
9bd97b82f43814f7c3d42f80f839ecf5  /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcsamtools.pxd
d8ed9b0dd171c009f2a9dabb80c22b27  /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcbcftools.cpython-39-x86_64-linux-gnu.so
837ad4bc90f057bf0cecf102edf4f6d5  /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcbcftools.pxd
```

Hope you see something!

**Comment by sbslee** [comment]
@jmarshall,

Since you mentioned that you were not able to reproduce the error with your Linux, I decided to test my local Linux box in addition to my linux server -- i.e. this is a different machine than the one I have been testing on:

```
$ uname -a
Linux localhost.localdomain 3.10.0-1160.36.2.el7.x86_64 #1 SMP Wed Jul 21 11:57:15 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```
Again, I installed the latest version (0.18.0) via conda:

```
$ conda create -n test pysam=0.18.0
```

I used the same testing script:

```
$ cat pysam_test.py 
import pysam
from pysam import bcftools
import faulthandler
print("version {} from {}".format(pysam.__version__, pysam))
faulthandler.enable()
bcftools.mpileup(
    'cram/sliced-bam01/NA06984.PyPGx-GRCh38.bam',
    '-f', 'Homo_sapiens_assembly38.no_alt.fasta'
)
```

Running the script gives the identical error:

```
$ python pysam_test.py 
version 0.18.0 from <module 'pysam' from '/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/__init__.py'>
Fatal Python error: Segmentation fault

Current thread 0x00007f1c20df0740 (most recent call first):
  File "/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/utils.py", line 56 in __call__
  File "/home/sbslee/Desktop/pysam_test.py", line 6 in <module>
Segmentation fault (core dumped)
```

I will also provide all the things you previously asked:

```
(test) [sbslee@localhost Desktop]$ LD_DEBUG=libs python pysam_test.py 2>&1 | grep libc...tools
     18986:	find library=libcsamtools.cpython-39-x86_64-linux-gnu.so [0]; searching
     18986:	  trying file=/home/sbslee/anaconda3/envs/test/lib/libcsamtools.cpython-39-x86_64-linux-gnu.so
     18986:	  trying file=/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcsamtools.cpython-39-x86_64-linux-gnu.so
     18986:	find library=libcbcftools.cpython-39-x86_64-linux-gnu.so [0]; searching
     18986:	  trying file=/home/sbslee/anaconda3/envs/test/lib/libcbcftools.cpython-39-x86_64-linux-gnu.so
     18986:	  trying file=/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcbcftools.cpython-39-x86_64-linux-gnu.so
     18986:	calling init: /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcbcftools.cpython-39-x86_64-linux-gnu.so
     18986:	calling init: /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcsamtools.cpython-39-x86_64-linux-gnu.so
```

```
(test) [sbslee@localhost Desktop]$ md5sum /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libc{utils,samtools,bcftools}*
200b3a84e14d88d8e69194ef1429cd12  /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcutils.cpython-39-x86_64-linux-gnu.so
8a907487f5a653e9409a1c66d0772d6e  /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcutils.pxd
f3617094dc72ea047ad4a30b4698f87d  /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcsamtools.cpython-39-x86_64-linux-gnu.so
9bd97b82f43814f7c3d42f80f839ecf5  /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcsamtools.pxd
d8ed9b0dd171c009f2a9dabb80c22b27  /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcbcftools.cpython-39-x86_64-linux-gnu.so
837ad4bc90f057bf0cecf102edf4f6d5  /home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/libcbcftools.pxd
```

My question to you: **When you did your testing, which did not reproduce the error, did you install 0.18.0 via conda?**

**Comment by jmarshall** [comment]
As you reported encountering the problem using pysam installed via conda, I have been attempting to reproduce it using pysam installed via conda. My testing was using pysam 0.18.0 via conda on CentOS 8.

I notice that you are using CentOS 7, so I have tried reproducing it via conda on a local CentOS 7 machine. I'm not able to reproduce any problem there either.

I guess it would also be useful to see the list of packages installed in your `test` environment (i.e., the output of `conda list`).

Is there any possibility of sshing into one of your affected machines myself, so I could poke at the affected pysam installation directly?

**Comment by sbslee** [comment]
Thanks for the answer. I just wanted to rule out the possibility that you were building the latest version from source, etc. And thanks for checking CentOS 7 vs. 8.

Following packages are found in the Linux server:

```
(test) [sbslee@bdcm03 ~]$ conda list
# packages in environment at /home/sbslee/anaconda3/envs/test:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
c-ares                    1.18.1               h7f98852_0    conda-forge
ca-certificates           2021.10.8            ha878542_0    conda-forge
krb5                      1.19.2               hcc1bbae_3    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
libcurl                   7.81.0               h2574ce0_0    conda-forge
libdeflate                1.9                  h7f98852_0    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 11.2.0              h1d223b6_12    conda-forge
libgomp                   11.2.0              h1d223b6_12    conda-forge
libnghttp2                1.46.0               h812cca2_0    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libssh2                   1.10.0               ha56f1ee_2    conda-forge
libstdcxx-ng              11.2.0              he4da1e4_12    conda-forge
libuuid                   2.32.1            h7f98852_1000    conda-forge
libzlib                   1.2.11            h36c2ea0_1013    conda-forge
ncurses                   6.3                  h9c3ff4c_0    conda-forge
openssl                   1.1.1l               h7f98852_0    conda-forge
pip                       22.0.3             pyhd8ed1ab_0    conda-forge
pysam                     0.18.0           py39h20405f9_0    bioconda
python                    3.9.10          h85951f9_2_cpython    conda-forge
python_abi                3.9                      2_cp39    conda-forge
readline                  8.1                  h46c0cb4_0    conda-forge
setuptools                60.9.0           py39hf3d152e_0    conda-forge
sqlite                    3.37.0               h9cd32fc_0    conda-forge
tk                        8.6.11               h27826a3_1    conda-forge
tzdata                    2021e                he74cb21_0    conda-forge
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
zlib                      1.2.11            h36c2ea0_1013    conda-forge
```

Following packages are found in my local Linux box:

```
(test) [sbslee@localhost cram]$ conda list
# packages in environment at /home/sbslee/anaconda3/envs/test:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
c-ares                    1.18.1               h7f98852_0    conda-forge
ca-certificates           2021.10.8            ha878542_0    conda-forge
krb5                      1.19.2               hcc1bbae_3    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
libcurl                   7.81.0               h2574ce0_0    conda-forge
libdeflate                1.9                  h7f98852_0    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 11.2.0              h1d223b6_12    conda-forge
libgomp                   11.2.0              h1d223b6_12    conda-forge
libnghttp2                1.46.0               h812cca2_0    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libssh2                   1.10.0               ha56f1ee_2    conda-forge
libstdcxx-ng              11.2.0              he4da1e4_12    conda-forge
libuuid                   2.32.1            h7f98852_1000    conda-forge
libzlib                   1.2.11            h36c2ea0_1013    conda-forge
ncurses                   6.3                  h9c3ff4c_0    conda-forge
openssl                   1.1.1l               h7f98852_0    conda-forge
pip                       22.0.3             pyhd8ed1ab_0    conda-forge
pysam                     0.18.0           py39h20405f9_0    bioconda
python                    3.9.10          h85951f9_2_cpython    conda-forge
python_abi                3.9                      2_cp39    conda-forge
readline                  8.1                  h46c0cb4_0    conda-forge
setuptools                60.9.1           py39hf3d152e_0    conda-forge
sqlite                    3.37.0               h9cd32fc_0    conda-forge
tk                        8.6.11               h27826a3_1    conda-forge
tzdata                    2021e                he74cb21_0    conda-forge
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
zlib                      1.2.11            h36c2ea0_1013    conda-forge
```

As for letting you ssh into my machines, I really wish I could, but these are secured machines so I'm afraid I can't. I know how frustrating it can be to remotely fix a problem for someone, so my sincere apology. But I promise I will be as responsive as possible on your requests.

Instead, in case the issue is related to my input files -- which is unlikely because I've tested multiple different files but just in case -- here are links to test files: [FASTA](https://1drv.ms/u/s!Apgoq3uQ2gCqgq03kaIEV-soQatrMA?e=F0zSDL) and [BAM](https://1drv.ms/u/s!Apgoq3uQ2gCqgq0r3d3QcNOFRwcMFQ?e=6xiU4h)

**Comment by sbslee** [comment]
@jmarshall,

> That's an identical list of packages to my environment that does not crash, so the cause doesn't seem to be related to that.

Thanks for checking the dependencies. So weird!

> This crash occurs before the files are opened — before it even gets into the actual bcftools code at all — so the contents of the test files are (almost certainly) irrelevant.

That's also been my impression.

As for testing whether the issue is specific to the `bcftools.mpileup` method, great question! Below is the results summary:

| System           | `bcftools.mpileup` | `bcftools.view` | `pysam.view` |
| ------------------ | ------------------------ | ------------------- | ------------------ |
| macOS           | O                           | O                     | O                   |
| Linux (server) | X                           | O                     | O                   |
| Linux (local)    | X                           | O                     | O                   |

I will attach the three scripts I used for above testing and their results from Linux (local) as representative example:

```
(test) [sbslee@localhost Desktop]$ cat test_bcftools_mpileup.py
import pysam
from pysam import bcftools
import faulthandler
print("version {} from {}".format(pysam.__version__, pysam))
faulthandler.enable()
results = bcftools.mpileup(
    'NA06984.PyPGx-GRCh38.sorted.bam',
    '-f', 'Homo_sapiens_assembly38.no_alt.fasta',
)
print(results)
(test) [sbslee@localhost Desktop]$ python test_bcftools_mpileup.py
version 0.18.0 from <module 'pysam' from '/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/__init__.py'>
Fatal Python error: Segmentation fault

Current thread 0x00007fc1ac28a740 (most recent call first):
  File "/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/utils.py", line 56 in __call__
  File "/home/sbslee/Desktop/test_bcftools_mpileup.py", line 6 in <module>
Segmentation fault (core dumped)
```

```
(test) [sbslee@localhost Desktop]$ cat test_bcftools_view.py
import pysam
from pysam import bcftools
import faulthandler
print("version {} from {}".format(pysam.__version__, pysam))
faulthandler.enable()
results = bcftools.view('ts1.merged.joint.filtered.vcf')
print(results[:100])
(test) [sbslee@localhost Desktop]$ python test_bcftools_view.py
version 0.18.0 from <module 'pysam' from '/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/__init__.py'>
##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##ALT=<ID=NON_REF,Descripti
```

```
(test) [sbslee@localhost Desktop]$ cat test_pysam_view.py
import pysam
from pysam import bcftools
import faulthandler
print("version {} from {}".format(pysam.__version__, pysam))
faulthandler.enable()
results = pysam.view('NA06984.PyPGx-GRCh38.sorted.bam')
print(results[:100])
(test) [sbslee@localhost Desktop]$ python test_pysam_view.py
version 0.18.0 from <module 'pysam' from '/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/__init__.py'>
A00296:33:HCN2LDSXX:3:1523:7012:7654	99	chr1	46795898	60	150M	=	46796151	403	CATTTATTGATTTGCATATGTTG
```

In conclusion, the issue does seem specific to the `bcftools.mpileup` method, and not the entire `bcftools` module, since `bcftools.view` works fine.

**Comment by jmarshall** [comment]
@sbslee: I have sent you updated copies of some of the `libc*.cpython-39-x86_64-linux-gnu.so` libraries from pysam 0.18.0, with extra debug logging in them. Please overwrite these three files in `/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam` (i.e., in your test environment, not a production one!) and rerun your `pysam_test.py` script.

If we are unlucky, these rebuilt files just will make the problem go away (and we will be none the wiser). But if we are lucky, you will get the same crash but it will be preceded by some `@JM@`-prefixed debugging output that will be instructive.

**Comment by sbslee** [comment]
@jmarshall,

Thank you very much sending the custom libraries! Below you can see the results. I guess we were lucky here...? I hope the debugging output will help you identify the problem!

```
(test) [sbslee@localhost Desktop]$ python test_bcftools_mpileup.py 
version 0.18.0 from <module 'pysam' from '/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/__init__.py'>
@JM@ In _pysam_dispatch()
@JM@ _pysam_dispatch() calling bcftools_dispatch()
@JM@ In bcftools_dispatch()
Fatal Python error: Segmentation fault

Current thread 0x00007fe6764e3740 (most recent call first):
  File "/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/utils.py", line 56 in __call__
  File "/home/sbslee/Desktop/test_bcftools_mpileup.py", line 6 in <module>
Segmentation fault (core dumped)
```

```
(test) [sbslee@localhost Desktop]$ python test_bcftools_view.py 
version 0.18.0 from <module 'pysam' from '/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/__init__.py'>
@JM@ In _pysam_dispatch()
@JM@ _pysam_dispatch() calling bcftools_dispatch()
@JM@ In bcftools_dispatch()
@JM@ _pysam_dispatch() after bcftools_dispatch
##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##ALT=<ID=NON_REF,Descripti
```

```
(test) [sbslee@localhost Desktop]$ python test_pysam_view.py 
version 0.18.0 from <module 'pysam' from '/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/__init__.py'>
@JM@ In _pysam_dispatch()
@JM@ _pysam_dispatch() calling samtools_dispatch()
@JM@ In samtools_dispatch()
@JM@ _pysam_dispatch() after samtools_dispatch
A00296:33:HCN2LDSXX:3:1523:7012:7654	99	chr1	46795898	60	150M	=	46796151	403	CATTTATTGATTTGCATATGTTG
```

**Comment by jmarshall** [comment]
That is indeed good news (!) that the problem persists. I've sent you another batch of custom libraries. With continued luck, we should be at the bottom of this with a total of four batches.

**Comment by sbslee** [comment]
I'm glad to hear things are moving! Below are the results from the new libraries:

```
(test) [sbslee@localhost Desktop]$ python test_bcftools_mpileup.py
version 0.18.0 from <module 'pysam' from '/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/__init__.py'>
@JM@ In _pysam_dispatch()
@JM@ _pysam_dispatch() calling bcftools_dispatch()
@JM@ In bcftools_dispatch() [patch files edition 2]
@JM@ In bcftools::main_mpileup()
@JM@ bcftools::main_mpileup(): post getopt
@JM@ bcftools::main_mpileup(): checking --fasta-ref
@JM@ bcftools::main_mpileup(): command line filenames
@JM@ In bcftools::mpileup()
Fatal Python error: Segmentation fault

Current thread 0x00007f8746515740 (most recent call first):
  File "/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/utils.py", line 56 in __call__
  File "/home/sbslee/Desktop/test_bcftools_mpileup.py", line 6 in <module>
Segmentation fault (core dumped)
```

```
(test) [sbslee@localhost Desktop]$ python test_bcftools_view.py
version 0.18.0 from <module 'pysam' from '/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/__init__.py'>
@JM@ In _pysam_dispatch()
@JM@ _pysam_dispatch() calling bcftools_dispatch()
@JM@ In bcftools_dispatch() [patch files edition 2]
@JM@ _pysam_dispatch() after bcftools_dispatch
##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##ALT=<ID=NON_REF,Descripti
```

```
(test) [sbslee@localhost Desktop]$ python test_pysam_view.py
version 0.18.0 from <module 'pysam' from '/home/sbslee/anaconda3/envs/test/lib/python3.9/site-packages/pysam/__init__.py'>
@JM@ In _pysam_dispatch()
@JM@ _pysam_dispatch() calling samtools_dispatch()
@JM@ In samtools_dispatch() [patch files edition 2]
@JM@ _pysam_dispatch() after samtools_dispatch
A00296:33:HCN2LDSXX:3:1523:7012:7654	99	chr1	46795898	60	150M	=	46796151	403	CATTTATTGATTTGCATATGTTG
```

**Comment by jmarshall** [comment]
That got further than I expected it to (“`In bcftools::mpileup()`”), which suggests that in fact it is getting as far as reading the input BAM file. And using the example file that you previously made available, I have been able to reproduce the problem.

It is similar to the previous report #928 in that the problem is separate functions in samtools and bcftools that happen to have the same name. Where previously the problem was the main `bam_mpileup()` entry point, this time it is various small utility functions that are used in both programs.

We will fix this properly for the next Pysam release. (I am drafting a PR to update to HTSlib et al 1.15, which will make this easier as many of these duplicate functions have disappeared with the removal of samtools's VCF/BCF-outputting mpileup functionality. The recent HTSlib et al release means that the next pysam release won't be too far off.)

In the meantime, I have submitted an update to the bioconda pysam package that should fix this problem. When PR bioconda/bioconda-recipes#33370 has been merged, you should be able to install a revised pysam-0.18.0-py3Xxyz_2 (note the `_2` at the end) package with a working `bcftools.mpileup()`.

Thanks for the bug report.

**Comment by sbslee** [comment]
@jmarshall,

I can't say thank you enough! I am aware the `bcftools` part of `pysam` is not as heavily used by users as the `samtools` part (i.e. low priority), so I really appreciate you putting in time and effort to fix these problems for a small group of people including myself.

Quick follow-up questions:

> And using the example file that you previously made available, I have been able to reproduce the problem.

Does this mean indeed the problem was input-specific?

> Where previously the problem was the main `bam_mpileup()` entry point, this time it is various small utility functions that are used in both programs.

Are you suggesting that these "small" utilities were malfunctioning in a OS-dependent manner (i.e. Linux vs. macOS)? If so, fascinating!

Again, thanks for everything, especially the PR. Looking forward to trying the new bioconda version.

**Comment by jmarshall** [comment]
I suspect what was critical about your input file is that it contains several readgroups, and in particular several readgroups for the same sample. The function causing the (first) crash was to do with sample handling. Unfortunately I had misinterpreted your initial comments about `in.bam` and `ref.fa` as meaning the input files were immaterial and didn't even need to exist, hence why I assumed the crash occurred earlier, before the input files were accessed.

It appears that POSIX does define what should happen if the address space contains multiple definitions of the same symbol, and for the purposes of what we see here it appears to be based on load order. I'm not going to look at that low-level CPython code to see how it loads these libraries, but it would be unsurprising if they were loaded in different orders on different OSes, or even on repeated installations in different directories on the same machine. I wouldn't even be surprised if it could potentially differ from run to run on the same machine due to [ASLR](https://en.wikipedia.org/wiki/Address_space_layout_randomization).

It's a vague enough area with enough moving parts (i.e., the unknown Python internals) that the only sensible approach is to avoid it altogether, by renaming the functions (within pysam) to avoid collisions.

**Comment by sbslee** [comment]
Thanks for the detailed explanations, much appreciated.

**Comment by jmarshall** [comment]
The fixed bioconda package is now available as pysam-0.18.0-py39h5030a8b_2 et al.

(But let's leave this issue open until it's fixed properly here in pysam.)

**Comment by sbslee** [comment]
@jmarshall,

I confirm installation of `pysam-0.18.0-py39h5030a8b_2` fixes the issue. Thanks!
